docs/installation/database/passing-configuration.rst

```diff
@@ -0,0 +1,430 @@
+
+The normal flow of configuration processing can be bypassed by `explicitly passing`_ in a complete configuration.
+Such explicit configuration blocks reading of configuration from environment variables.
+
+Otherwise, configuration is read from a `configuration file`_ and/or `environment variables`_.
+
+One configuration can define multiple environments, so users must `choose one`_.
+
+The configuration engine in 1.9 is not 100% compatible with the previous configuration engine.  Advanced
+users and developers upgrading 1.8 systems should read the `migration notes`_.
+
+.. _`explicitly passing`: #Explicit-configurations
+.. _`configuration file`: #File-configurations
+.. _`choose one`: #The-Active-Environment
+.. _`environment variables`: #Generic-Environment-Variable-Overrides
+.. _`migration notes`: #Migrating-from-datacube-1.8
+
+1. Explicit configuration
+-------------------------
+
+Configuration can be passed in explicitly, without ever reading from a configuration file on disk.
+
+When explicit configuration is passed in, it takes precedence over configuration by environment variable.
+
+1a. Via Python (str or dict)
+++++++++++++++++++++++++++++
+
+A valid configuration dictionary can be passed in directly to the
+:py:class:`Datacube` constructor with the ``raw_config`` argument, without
+serialising to a string:
+
+.. code-block:: python
+
+   dc = Datacube(raw_config={
+      "default": {
+         "index_driver": "postgres",
+         "db_url": "postgresql:///mydb"
+      }
+   })
+
+The ``raw_config`` argument can also be passed config as a string, in either INI or YAML format:
+
+.. code-block:: python
+
+   dc = Datacube(raw_config="""
+   default:
+     # Connect to database mydb over local socket with OS authentication.
+     db_url: postgresql:///mydb
+   """)
+
+1b. As a string, via the datacube CLI
++++++++++++++++++++++++++++++++++++++
+
+The contents of a configuration file can be passed into the ``datacube`` CLI via the ``-R`` or
+``--raw-config`` command line option:
+
+::
+
+   datacube --raw-config "default: {db_database: this_db}"
+
+Output from a script that generates a configuration file dynamically can be passed in using
+a BASH backquote string:
+
+::
+
+   datacube --raw-config "`config_file_generator --option blah`"
+
+1c. As a string, via an Environment Variable
+++++++++++++++++++++++++++++++++++++++++++++
+
+If raw configuration has not been passed in explicitly via methods 1a. or 1b.
+above, then the contents of a configuration file can be written in full to the
+:envvar:`ODC_CONFIG` environment variable:
+
+.. code-block:: console
+
+   $ ODC_CONFIG="default: {db_database: this_db}"
+   $ datacube check    # will use the this_db database
+
+
+2. File configuration
+---------------------
+
+.. highlight:: python
+
+If explicit configuration has not been passed in, ODC attempts to find a configuration file.
+
+Only one configuration file is read.
+
+This is different to previous versions of the Open Data Cube,
```

Comment:
Should this and the next paragraph perhaps be moved to a migration guide (that should probably contain other things relevant for someone upgrading from 1.8l)?

Reply:
There's configuration migration notes at the end of this section already.

There's also a draft of the migration notes sitting there - will be integrated better into the docs in another PR.